### PR TITLE
Census services

### PR DESCRIPTION
**Pull request by mirunaish**
- tree census
	- create: only if plot census is active, user is assigned to plot census, tree has not been censused already
	- edit: only if plot census is active, user is assigned to plot census
- plot census
	- create: only if forest census is active, plot has not been censused already
	- submitForReview: if census is in progress, all trees have been censused
- plot census assignment
	- create: if plot is in progress / does not exist, user is not already assigned
- forest census
	- create: if an active forest census does not already exist
	- close: if forest census is open, all plots have been censused

### CODE REVIEW

File: packages/server/src/services/forest-census-service.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+export const closeForestCensus = async (params: { forestId: string }) => {
+  const { forestId } = params;
+
+  // find active forest census
+  const activeCensuses = await ForestCensusModel.findAll({
+    where: { forestId: { [Op.eq]: forestId }, active: { [Op.eq]: true } },
+  });
+  if (activeCensuses.length > 1) {
+    throw Error("Error: more than one active census");
```
Comment (mirunaish) [inline packages/server/src/services/forest-census-service.ts:30]: i ran into this situation on multiple occasions: there's something in the database that is supposed to be unique, but there's no formal constraints forcing it to be so. the routes are designed to catch and disallow duplicates, but if the admins change something through retool it could introduce duplicates. is it necessary to findAll and ensure the thing is unique, or is it enough to findOne and assume it is unique?

File: packages/server/src/services/plot-census-assignment-service.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  // this should never happen
+  if (existingCensus.length > 1) {
+    throw Error(
+      "Fatal error: more than one open census on this plot. Ask an administrator for assistance"
+    );
+  }
+
+  if (existingCensus.length > 0) {
+    // if one open census, check whether it is pending or in progress
+    if (existingCensus[0].status == PlotCensusStatuses.InProgress) {
```
Comment (mirunaish) [inline packages/server/src/services/plot-census-assignment-service.ts:34]: currently plotCensusAssignmentService checks whether a plot census exists and if it is in progress or pending, and the createPlotCensus function in plotCensusService checks whether an approved census exists in this census generation/forest census (and throws an error if it does, which the createPlotCensusAssignment function passes on to the caller). the createPlotCensus function also checks for in progress/pending plot censuses, so it can function independently of plotCensusAssignment. all the conditions got pretty confusing. is it better to have all the conditions in both places, or try to keep the conditions in createPlotCensus and have createPlotCensusAssignment pass on the errors, or keep this middle ground?

File: packages/server/src/services/plot-census-service.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
+import { PlotCensus, PlotCensusStatuses } from "@ong-forestry/schema";
+import PlotCensusModel from "db/models/plot-census";
+import PlotModel from "db/models/plot";
+import TreeCensusModel from "db/models/tree-census";
+import TreeModel from "db/models/tree";
+import ForestCensusModel from "db/models/forest-census";
+import { Op } from "sequelize";
+
+const uuid = require("uuid4");
+
```
Comment (mirunaish) [inline packages/server/src/services/plot-census-service.ts:10]: i think plot-census and tree-census may need get routes so the app can know what color to display each plot/tree in, but i'm not sure what those routes would have to look like or how displaying tree census information on the map / in the form will work

File: packages/server/src/services/tree-census-service.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+const constructQuery = (params: EditTreeCensusParams) => {
+  const { treeTags } = params;
+  const query: any = { where: {} };
+  if (treeTags) {
+    query.where.treeTags = { [Op.in]: treeTags };
+  }
+  return query;
+};
+
+export const editTreeCensuses = async (
```
Comment (mirunaish) [inline packages/server/src/services/tree-census-service.ts:83]: what happens if a user creates a tree census and another user edits it? which one of them is the author of that tree census? both?

File: packages/server/src/services/tree-census-service.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    throw Error("This tree has already been censused.");
+  }
+
+  return await TreeCensusModel.create({
+    ...treeCensus,
+    plotCensusId,
+  });
+};
+
+export interface EditTreeCensusParams {
```
Comment (mirunaish) [inline packages/server/src/services/tree-census-service.ts:70]: i know these don't really follow the same structure as the other services. i was thinking about the edit(/get/delete) query parameters and i think that in most cases the only attribute that would be realistically used to select a row to be edited is the primary key. i'm not sure including any of the other attributes is necessary

File: packages/server/src/services/forest-census-service.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        );
+      }
+    })
+  );
+
+  return await ForestCensusModel.update(
+    { active: false },
+    { where: { id: { [Op.eq]: activeCensuses[0].id } } }
+  );
+};
```
Comment (mirunaish) [inline packages/server/src/services/forest-census-service.ts:62]: is there a way to allow renaming a forest census with sql directly from retool but make the active status only changeable by the route?

Comment (ziruihao) [reply packages/server/src/services/forest-census-service.ts:62]: We can add Retool SQL queries to rename forest census without affecting the active status!

**Comment by mirunaish** [comment]
i should mention that i haven't had a chance to test any of the new routes yet. i have tested the plot id migration and it works as far as i can tell

File: packages/server/src/services/plot-census-assignment-service.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    },
+  });
+
+  // there should only be one un-approved census per plot
+  // this should never happen
+  if (existingCensus.length > 1) {
+    throw Error(
+      "Fatal error: more than one open census on this plot. Ask an administrator for assistance"
+    );
+  }
```
Comment (ziruihao) [inline packages/server/src/services/plot-census-assignment-service.ts:30]: Nice — still a good condition to catch!

File: packages/server/src/services/plot-census-service.ts (hunk 3 above)
Comment (ziruihao) [reply packages/server/src/services/plot-census-service.ts:10]: Good point, querying routes will be necessary!

We'll start with `forest_census`:
- Get forest census with params: `forestId`

Then with `plot_census`:
- Get plot census with params: `forestCensusId`, `plotId`

Then for `tree_census`:
- Get tree census with params: `plotCensusId`, `treeId`, `authorId`

File: packages/server/src/services/tree-census-service.ts (hunk 5 above)
Comment (ziruihao) [reply packages/server/src/services/tree-census-service.ts:70]: I see what you mean, but my take is that it's easier to have just one params interface, instead of creating two param interfaces (one for GET, other for UPDATE / DELETE).

I think it's OK for that one params interface to be overloaded with many querying options, even fi they're not used all the time.

File: packages/server/src/services/tree-census-service.ts (hunk 4 above)
Comment (ziruihao) [reply packages/server/src/services/tree-census-service.ts:83]: Last updated person :)

File: packages/server/src/services/plot-census-assignment-service.ts (hunk 2 above)
Comment (ziruihao) [reply packages/server/src/services/plot-census-assignment-service.ts:34]: Some really good thoughts about code sustainability! 🌱

A clean way to think about this is the input-output (IO) states of each function. The IO of `createPlotCensus` is that we  _ensure_ that the plot census exists, irregardless of whether it already existed.
- If not preexisting, create one.
- If preexisting, throw a unique error.

The IO of `createPlotCensusAssignment` is to start in a state where you don't care if there's already a plot census or a plot census assignment, and end in a state where you're _certain_ that there's a plot census _and_ a plot census assignment.

So, the **slimmest** code is:
- `createPlotCensus` contains all necessary prechecks
- `createPlotCensusAssignment` calls `createPlotCensus` first, catching ONLY the error thrown if there already existed a plot census (don't catch other errors!), which puts us in a state _ensuring_ a preexisting plot census. Then we just have to add a new plot census assignment

File: packages/server/src/services/forest-census-service.ts (hunk 1 above)
Comment (ziruihao) [reply packages/server/src/services/forest-census-service.ts:30]: Yeah because we don't have a DB-layer check on tag duplication, we can't assume tags are unique at the app-layer anymore. We can actually disallow Retool to modify the tag, but we still won't have a _guarantee_. With regards for `findOne` or `findAll`, we should aim to always query stuff by their PKs, which lets us use `findOne` with confidence.

**Review by ziruihao — COMMENTED** [review]
Great progress so far! Just some responses to comments.

File: packages/server/src/services/plot-census-service.ts [hunk 8]
```diff
@@ -67,6 +74,74 @@ export const createPlotCensus = async (plotId: string) => {
...
+    );
+  }
+
+  // need to handle limit and offset manually
+  if (params.offset) {
+    plotCensuses = plotCensuses.filter(
+      (plotCensus, index) => index >= (params.offset ?? 0)
+    );
+  }
+  if (params.limit) {
```
Comment (mirunaish) [inline packages/server/src/services/plot-census-service.ts:136]: would limit ever be 0?

Comment (ziruihao) [reply packages/server/src/services/plot-census-service.ts:125]: Good point, but for simplicity sake, we can just assume that if a person wants `limit = 0` they can just not call this API :)

Comment (ziruihao) [reply packages/server/src/services/plot-census-service.ts:125]: I do wish JS had an elegant "check if defined" method, aside from `typeof var !== 'undefined && var != null`

File: packages/server/src/services/forest-census-service.ts [hunk 9]
```diff
@@ -19,6 +19,36 @@ export const createForestCensus = async (forestCensus: ForestCensus) => {
...
+    query.where.forestId = {
+      [Op.eq]: forestId,
+    };
+  }
+  if (limit) {
+    query.limit = limit;
+  }
+  if (offset) {
+    query.offset = offset;
+  }
```
Comment (ziruihao) [inline packages/server/src/services/forest-census-service.ts:46]: I believe we can just initialize `query` as

`query = {
    where: {},
    limit,
    offset
}`

to begin with, so we can save these lines of code.

File: packages/server/src/services/plot-census-service.ts [hunk 10]
```diff
@@ -73,6 +74,74 @@ export const createPlotCensus = async (plotId: string) => {
...
+      })
+    ).map((plotCensusAssignment) => plotCensusAssignment.plotCensusId);
+
+    // intersect found plot censuses with plot censuses assigned to this user
+    plotCensuses = plotCensuses.filter((plotCensus) =>
+      userCensuses.includes(plotCensus.id)
+    );
+  }
+
+  // need to handle limit and offset manually
```
Comment (ziruihao) [inline packages/server/src/services/plot-census-service.ts:119]: What's the reason for needing to handle limit and offsets manually here?

File: packages/server/src/db/migrations/20220429051535-change-tree-pk.js [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+module.exports = {
+  async up(queryInterface, Sequelize) {
+    const transaction = await queryInterface.sequelize.transaction();
+    try {
+      // add new id column
+      await queryInterface.addColumn(
+        "trees",
+        "id",
+        { type: Sequelize.UUID, default: Sequelize.UUID, unique: true },
```
Comment (ziruihao) [inline packages/server/src/db/migrations/20220429051535-change-tree-pk.js]: If this works that's totally good, but if it's giving you trouble, we use:

`defaultValue: Sequelize.UUIDV4`

File: packages/server/src/db/migrations/20220429051535-change-tree-pk.js [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+            "tree_census",
+            { treeId: dictionary[treeTag] },
+            { treeTag },
+            { transaction }
+          );
+        })
+      );
+      // move foreign key constraint to new column
+      await queryInterface.sequelize.query(
+        'ALTER TABLE tree_census DROP CONSTRAINT "tree_census_treeTag_fkey"; ALTER TABLE tree_census ADD CONSTRAINT "tree_census_treeId_fkey" FOREIGN KEY ("treeId") REFERENCES trees (id)',
```
Comment (ziruihao) [inline packages/server/src/db/migrations/20220429051535-change-tree-pk.js:58]: Also add the `NOT NULL` constraint so that `treeId` has to be defined in `tree_census` table! Overall neat migration!

File: packages/server/src/db/models/tree.ts [hunk 13]
```diff
@@ -35,6 +36,10 @@ class Tree
   implements Omit<ITree, "createdAt" | "updatedAt">
 {
   @PrimaryKey
+  @Default(DataTypes.UUID)
```
Comment (ziruihao) [inline packages/server/src/db/models/tree.ts:41]: If this gives errors, try `@Default(DataTypes.UUIDV4)`

File: packages/server/src/services/tree-service.ts [hunk 14]
```diff
@@ -1,12 +1,36 @@
 import { Tree } from "@ong-forestry/schema";
 import TreeModel from "db/models/tree";
+import PlotModel from "db/models/plot";
 import { Op } from "sequelize";
 
-export const createTrees = async (tree: Tree) => {
+export const createTree = async (tree: Tree) => {
+  // ensure tag unique in this forest
```
Comment (ziruihao) [inline packages/server/src/services/tree-service.ts:8]: Awesome — thanks for adding this!

File: packages/server/src/routes/tree-routes.ts [hunk 15]
```diff
@@ -32,10 +32,6 @@ const parseParams = (query: any) => ({
   plotXMax: parseFloat(query.plotXMax as string),
   plotYMin: parseFloat(query.plotYMin as string),
   plotYMax: parseFloat(query.plotYMax as string),
-  dbhMin: parseFloat(query.dbhMin as string),
```
Comment (ziruihao) [inline packages/server/src/routes/tree-routes.ts:42]: Oops I missed this from before — thanks for removing!

File: packages/server/src/db/migrations/20220430023414-add-flagged-column.js [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+  async up(queryInterface, Sequelize) {
+    const transaction = await queryInterface.sequelize.transaction();
+
+    try {
+      await queryInterface.addColumn(
+        "tree_census",
+        "flagged",
+        {
+          type: Sequelize.BOOLEAN,
+          defaultValue: false,
```
Comment (ziruihao) [inline packages/server/src/db/migrations/20220430023414-add-flagged-column.js:16]: Wouldn't hurt to also have `allowNull: false`: https://stackoverflow.com/a/11862246

File: packages/server/src/routes/membership-routes.ts [hunk 17]
```diff
@@ -10,7 +10,7 @@ import { requireAuth } from "services/auth-service";
 
 const membershipRouter = express.Router();
 
-membershipRouter.post<{}, any, Membership>(
+membershipRouter.post<{}, any, { email: string; teamId: string }>(
```
Comment (ziruihao) [inline packages/server/src/routes/membership-routes.ts:13]: We should do `Pick<Membership, 'email' | 'teamId'>`. The only reason is for code coherence. If the `Membership` interface evolves in the future, `{ email: string; teamId: string}` is wholly separate from it, and doesn't know to evolve along with `Membership`. The `Pick` operator will tie these together.

**Review by ziruihao — APPROVED** [review]
Tons of great careful work! Some comments but otherwise good to call from those!

File: packages/server/src/services/plot-census-service.ts (hunk 10 above)
Comment (mirunaish) [reply packages/server/src/services/plot-census-service.ts:119]: because i manually intersect the results of two queries - if both queries return `limit` elements but they don't coincide, the final result will have fewer than `limit` elements

File: packages/server/src/db/migrations/20220429051535-change-tree-pk.js (hunk 11 above)
Comment (mirunaish) [reply packages/server/src/db/migrations/20220429051535-change-tree-pk.js]: it does work, but i can change it

File: packages/server/src/routes/membership-routes.ts (hunk 17 above)
Comment (mirunaish) [reply packages/server/src/routes/membership-routes.ts:13]: will this work if Membership only contains userId and teamId?

File: packages/server/src/services/forest-census-service.ts (hunk 9 above)
Comment (mirunaish) [reply packages/server/src/services/forest-census-service.ts]: this doesn't seem to work. if limit and offset are not defined the generated SQL query ends in "LIMIT NaN OFFSET NaN" which breaks

File: packages/server/src/services/forest-census-service.ts [hunk 18]
```diff
@@ -55,6 +55,10 @@ export const getForestCensuses = async (params: GetForestCensusesParams) => {
 export const closeForestCensus = async (params: { forestId: string }) => {
   const { forestId } = params;
 
+  if (forestId == null) {
+    throw new Error("You must specify a forest.");
+  }
+
```
Comment (mirunaish) [inline packages/server/src/services/forest-census-service.ts:61]: is there a better way to ensure a query parameter is not null? or can we assume that the frontend will always send valid requests?

File: packages/server/src/services/forest-census-service.ts [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+  if (activeCensuses.length > 1) {
+    throw new Error("Error: more than one active census");
+  }
+  if (activeCensuses.length < 1) {
+    throw new Error("There is no active forest census.");
+  }
+
+  // ensure all plots in this forest have approved plot censuses
+  // select all plots in the forest
+  const plots = await getPlots({ forestId });
```
Comment (mirunaish) [inline packages/server/src/services/forest-census-service.ts:76]: is it necessary to test for unusual edge cases such as a forest with no plots?

File: packages/server/src/services/forest-census-service.ts (hunk 9 above)
Comment (ziruihao) [reply packages/server/src/services/forest-census-service.ts:47]: gotcha good to know! sorry hahaha I didn't realize that, thanks for experimenting!

File: packages/server/src/services/plot-census-service.ts (hunk 10 above)
Comment (ziruihao) [reply packages/server/src/services/plot-census-service.ts:138]: ah smart smart

File: packages/server/src/routes/membership-routes.ts (hunk 17 above)
Comment (ziruihao) [reply packages/server/src/routes/membership-routes.ts]: Oh I see what this route is doing — nevermind sorry my comment isn't applicable!

Comment (ziruihao) [reply packages/server/src/routes/membership-routes.ts]: _Ideally_ we'd do like `Pick<User | 'email'>` `Pick<Membership | 'teamId'>, but like that's so much work and not really necessary, because we'd likely **never** change the `email` or `teamId` fields on these schemas.

File: packages/server/src/services/forest-census-service.ts (hunk 18 above)
Comment (ziruihao) [reply packages/server/src/services/forest-census-service.ts]: We can create a middleware that validates this, but no we can't assume the frontend will comply :(

File: packages/server/src/services/forest-census-service.ts (hunk 19 above)
Comment (ziruihao) [reply packages/server/src/services/forest-census-service.ts:83]: Yes, good call!

File: packages/server/src/services/plot-census-service.ts [hunk 20]
```diff
@@ -12,7 +12,9 @@ import {
 
 const uuid = require("uuid4");
 
-export const createPlotCensus = async (plotId: string) => {
+export const createPlotCensus = async (params: Pick<PlotCensus, "plotId">) => {
+  const { plotId } = params;
+
```
Comment (ziruihao) [inline packages/server/src/services/plot-census-service.ts:17]: 10/10 for adopting this practice everywhere!

it'll make our lives so much easier if we change our schemas, because TS will flag the changes

**Review by ziruihao — APPROVED** [review]
Huge for the quick turnaround on these changes! I appreciate the meticulous effort with all these cases.
